### PR TITLE
Multiple hostnames

### PR DESCRIPTION
**Pull request by dr107**
Change the cli to accept multiple hostnames on the command line.

TODO:
	add tests
	update documentation

### CODE REVIEW

**Comment by grittygrease** [comment]
I'm not sure this is correct. It seems to change the behavior in the local signer tests:
FAIL	github.com/cloudflare/cfssl/signer/local	2.034s

**Comment by lziest** [comment]
A rebase may clear the coverall failure.

File: api/api_signer.go [hunk 1]
```diff
@@ -73,6 +73,33 @@ func NewSignHandlerFromSigner(signer signer.Signer) (h HTTPHandler, err error) {
 	}, nil
 }
 
+// a type meant to be unmarshalled from JSON so that there can be a
+// hostname field in the API
+type jsonSignRequest struct {
```
Comment (lziest) [inline api/api_signer.go]: This seems unnecessary. But I understand it is critical to change the API look and feel. Leave it for now but add a TODO comment.

File: signer/local/local_test.go [hunk 2]
```diff
@@ -489,7 +489,12 @@ func TestOverrideSubject(t *testing.T) {
 
 	s := newCustomSigner(t, testECDSACaFile, testECDSACaKeyFile)
 
-	certPEM, err := s.Sign(signer.SignRequest{Hostname: "localhost", Request: string(csrPEM), Subject: req})
+	certPEM, err := s.Sign(signer.SignRequest{
+		Hosts:   []string{"127.0.0.1", "localhost"},
+		Request: string(csrPEM),
+		Subject: req,
+	})
+
```
Comment (lziest) [inline signer/local/local_test.go]: Consider add a tests block below to verify the hosts are present in the right place.

File: signer/signer.go [hunk 3]
```diff
@@ -112,7 +118,7 @@ func DefaultSigAlgo(priv crypto.Signer) x509.SignatureAlgorithm {
 // builds a certificate template from it. If not nil, the subject
 // information from subject will be used in place of the information in
 // the CSR.
```
Comment (lziest) [inline signer/signer.go]: Add documentation for hosts.

File: signer/local/local.go [hunk 4]
```diff
@@ -133,26 +133,18 @@ func (s *Signer) Sign(req signer.SignRequest) (cert []byte, err error) {
 			cferr.BadRequest, errors.New("not a certificate or csr"))
 	}
 
-	template, err := signer.ParseCertificateRequest(s, block.Bytes, req.Subject)
+	template, err := signer.ParseCertificateRequest(s, block.Bytes, req.Subject, req.Hosts)
```
Comment (lziest) [inline signer/local/local.go]: ParseCertificateRequest already did the hosts appending, remove the duplicate code below.

File: api/api_signer.go [hunk 5]
```diff
@@ -73,6 +73,33 @@ func NewSignHandlerFromSigner(signer signer.Signer) (h HTTPHandler, err error) {
 	}, nil
 }
 
+// a type meant to be unmarshalled from JSON so that there can be a
```
Comment (kisom) [inline api/api_signer.go]: A nit, but can you make this a complete sentence? It'll be easier to read and understand later on.

File: api/api_signer.go (hunk 1 above)
Comment (dr107) [inline api/api_signer.go]: As in TODO: change the API to make this unnecessary?

Comment (lziest) [inline api/api_signer.go]: yes

File: initca/initca_test.go [hunk 6]
```diff
@@ -146,11 +146,10 @@ func TestInitCA(t *testing.T) {
...
 			req := signer.SignRequest{
-				Hostname: hostname,
-				Request:  string(csrBytes),
-				Subject:  nil,
-				Profile:  "",
-				Label:    ""}
+				Request: string(csrBytes),
+				Hosts:   signer.SplitHosts(hostname),
+				Profile: "",
+				Label:   ""}
```
Comment (kisom) [inline initca/initca_test.go]: This brace should be on a new line.

File: signer/local/local_test.go (hunk 2 above)
Comment (dr107) [inline signer/local/local_test.go]: How should I parse certPEM? I tried using x509.ParseCertificate, but when I do I get

asn1: structure error: tags don't match (16 vs {class:0 tag:13 length:45 isCompound:true}) {optional:false explicit:false application:false defaultValue:<nil> tag:<nil> stringType:0 set:false omitEmpty:false} certificate @2

Is there some CFSSL function I should be using?

Comment (kisom) [inline signer/local/local_test.go]: helpers.ParseCertificatePEM should do the trick.

Comment (lziest) [inline signer/local/local_test.go]: it is parsed a few lines down below

**Comment by lziest** [comment]
LGTM, please rebase to the latest master.

**Comment by kisom** [comment]
*Note*: build failure is known issue pushing to coveralls from a forked repo. All tests pass; `gofmt` and `golint` also have no errors.

LGTM.

**Pull request merged**